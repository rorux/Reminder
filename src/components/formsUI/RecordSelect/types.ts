export type TSelectParams = {
  value: string | number;
  name: string | number;
};

export type TSelectProps = {
  name: string;
  params: Array<TSelectParams>;
  setter: any;
  value?: any;
};
