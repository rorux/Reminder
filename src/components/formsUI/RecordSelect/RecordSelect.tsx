import React, { useEffect } from 'react';
import Box from '@mui/material/Box';
import InputLabel from '@mui/material/InputLabel';
import MenuItem from '@mui/material/MenuItem';
import FormControl from '@mui/material/FormControl';
import Select, { SelectChangeEvent } from '@mui/material/Select';
import { TSelectProps } from './types';

const MenuProps = {
  PaperProps: {
    style: {
      maxHeight: 170,
    },
  },
};

const RecordSelect: React.FC<TSelectProps> = ({ name, params, setter, value }) => {
  const [selectValue, setSelectValue] = React.useState('');

  useEffect(() => {
    if (value) setSelectValue(value);
  }, [value]);

  const handleChange = (event?: SelectChangeEvent) => {
    const value = event?.target.value as string;
    setSelectValue(value);
    setter(value);
  };

  return (
    <Box>
      <FormControl fullWidth required>
        <InputLabel id="select-label">{name}</InputLabel>
        <Select
          labelId="select-label"
          id="select"
          value={selectValue}
          label={name}
          onChange={handleChange}
          MenuProps={MenuProps}
        >
          {params.map((item) => {
            return (
              <MenuItem value={item.value} key={item.value}>
                {item.name}
              </MenuItem>
            );
          })}
        </Select>
      </FormControl>
    </Box>
  );
};

export default RecordSelect;
