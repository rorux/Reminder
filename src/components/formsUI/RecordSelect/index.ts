import RecordSelect from './RecordSelect';

export default RecordSelect;
