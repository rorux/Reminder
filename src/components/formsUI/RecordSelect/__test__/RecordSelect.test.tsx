import React from 'react';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import Enzyme, { shallow } from 'enzyme';
import { Select } from '@mui/material';
import RecordSelect from '../RecordSelect';

Enzyme.configure({ adapter: new Adapter() });

const setterFunc = jest.fn();

describe('RecordSelect component', () => {
  let component: Enzyme.ShallowWrapper;
  let useEffect: any;
  let setState: any;
  let useStateSpy;

  const mockUseEffect = () => {
    useEffect.mockImplementationOnce((f: any) => f());
  };

  beforeEach(() => {
    useEffect = jest.spyOn(React, 'useEffect');
    setState = jest.fn();
    useStateSpy = jest.spyOn(React, 'useState');
    // @ts-ignore
    useStateSpy.mockImplementation((init) => [init, setState]);
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('Should render RecordSelect component', () => {
    component = shallow(
      <RecordSelect
        name="name"
        params={[{ value: 'value', name: 'name' }]}
        setter={setterFunc}
        value="value"
      />
    );
    expect(component).toMatchSnapshot();
  });

  it('Should be called "useState"', () => {
    mockUseEffect();
    component = shallow(
      <RecordSelect
        name="name"
        params={[{ value: 'value', name: 'name' }]}
        setter={setterFunc}
        value="value"
      />
    );
    expect(setState).toHaveBeenCalledWith('value');
  });

  it('handleChange should call setState and setter func', () => {
    component = shallow(
      <RecordSelect
        name="name"
        params={[{ value: 'value', name: 'name' }]}
        setter={setterFunc}
        value="value"
      />
    );
    // @ts-ignore
    component.find(Select).props().onChange();
    expect(setState).toHaveBeenCalled();
    expect(setterFunc).toHaveBeenCalled();
  });
});
