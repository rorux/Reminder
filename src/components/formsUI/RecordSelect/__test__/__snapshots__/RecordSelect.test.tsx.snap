// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecordSelect component Should render RecordSelect component 1`] = `
<ForwardRef(Box)>
  <ForwardRef(FormControl)
    fullWidth={true}
    required={true}
  >
    <ForwardRef(InputLabel)
      id="select-label"
    >
      name
    </ForwardRef(InputLabel)>
    <ForwardRef(Select)
      MenuProps={
        Object {
          "PaperProps": Object {
            "style": Object {
              "maxHeight": 170,
            },
          },
        }
      }
      id="select"
      label="name"
      labelId="select-label"
      onChange={[Function]}
      value=""
    >
      <ForwardRef(MenuItem)
        key="value"
        value="value"
      >
        name
      </ForwardRef(MenuItem)>
    </ForwardRef(Select)>
  </ForwardRef(FormControl)>
</ForwardRef(Box)>
`;
